[BUG][java][jersey2] Generated java jersey2 client fails compilation when 'allOf' child has enum constraint

#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] What's the version of OpenAPI Generator used?
- [x] Have you search for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Bounty to sponsor the fix ([example](https://www.bountysource.com/issues/66123212-javascript-client-produces-a-wrong-object-for-a-string-enum-type-that-is-used-with-ref))

<!--
Please follow the issue template below for bug reports.
Also please indicate in the issue title which language/library is concerned. Eg:  [BUG][JAVA] Bug generating foo with bar 
-->

##### Description

The generated Java client code with the jersey2 library fails compilation when:
1. A **Animal** schema is defined with a property **P1** of type "string".
1. A **Cat** child schema is defined with 'allOf' Animal.
1. The **P1** property is redefined in the **Cat** child class.
1. A enum constraint is added to property **P1** in the child class.

The same problem is likely to occur with 'oneOf' and 'anyOf', though I have not tested these particular scenarios.

##### openapi-generator version

openapi master branch as of August 19th 2020

##### OpenAPI declaration file content or url

Here is a minimal OpenAPI doc to reproduce the problem. The full OpenAPI doc is available in PR #7251 at https://github.com/OpenAPITools/openapi-generator/pull/7251/files#diff-aa56ad872f3213179c27da9d73dec49aR2090

```yaml
    Animal:
      type: object
      properties:
       # Define a pet_type property of type "string" with no enum constraint.
       # It's just a plain string with no enum constraint.
        pet_type:
          type: string
    Cat:
      allOf:
        - $ref: '#/components/schemas/Animal'
        - type: object
          properties:
            pet_type:
              # Add a 'enum' constraint to the pet_type property and a 'default' value.
              # This is useful because it specifies there is only one possible value for this schema.
              # The code generator should be able to leverage this piece of information to generate
              # a Java class constructor where the 'pet_type' property is initialized to 'Cat' by default.
              # Without the enum constraint, there is no way for the code generator to determine
              # what are the possible values of 'pet_type'.
              # Without the enum, the developer creating a 'Cat' object using Java would be forced
              # to explicitly set the value of 'pet_type'. That would be both surprising and tedious
              # because humans know there is only one possible value.
              enum:
              - Cat
              default: Cat
```

##### Generation Details

<!-- 
    Prefer CLI steps, including the language, libraries and various options. 
    Providing config-based settings allows for simpler testing across CLI and plugins. 
    For examples, see https://github.com/OpenAPITools/openapi-generator/tree/master/bin/configs
-->

##### Steps to reproduce

1. Checkout branch from PR #7215.
1. Type
```bash
bin/generate-samples.sh bin/configs/java-jersey2-8.yaml
```

##### Related issues/PRs

I have raised PR #7251 to show how the generated code is failing.

##### Suggest a fix

First, let's look at the problems:
1. A field with the same name is redefined in Java sub-classes.
1. Getters and Setters methods are overridden in the Java sub-classes but with a different signature
1. There is no `@Override` annotation in the generated code.
1. More generally the code generator is not aware that methods are overridden in sub-classes.

The code for the parent and child class is generated as shown below.

```java
public class Animal {
  public String getPetType() {
    return petType;
  }
```

```java
public class Cat extends Animal {
  // Overrides getPetType from super class, but the return value is not compatible.
  public PetTypeEnum getPetType() {
    return petType;
  }
```

This causes the following compilation error:
```bash
Compilation failure
[ERROR] error: getPetType() in Cat cannot override getPetType() in Animal
[ERROR] return type PetTypeEnum is not compatible with String
```

Proposed changes:
1. Detect when property is redefined in 'allOf'/'oneOf'/'anyOf' child.
1. Detect which generated methods are overridden in java subclasses, e.g. `getPetType` is overridden. Ensure the generated methods have the same signature.
1. Do not redefine field in sub-class. Java fields cannot be overridden in a subclass. If a field is redefined in a subclass with the same name as a field in the superclass, the field in the subclass will hide (shadow) the field in the superclass. If the subclass tries to access the field, it will access the field in the subclass. However, if the subclass calls up into a method in the superclass, and that method accesses the field with the same name as in the subclass, it is the field in the superclass that is accessed.

--- a/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/JavaClientCodegen.java
+++ b/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/JavaClientCodegen.java
@@ -784,7 +784,32 @@ public class JavaClientCodegen extends AbstractJavaCodegen
                     // only add JsonNullable and related imports to optional and nullable values
                     addImports |= isOptionalNullable;
                     var.getVendorExtensions().put("x-is-jackson-optional-nullable", isOptionalNullable);
+
+                    if (Boolean.TRUE.equals(var.getVendorExtensions().get("x-enum-as-string"))) {
+                        // treat enum string as just string
+                        var.datatypeWithEnum = var.dataType;
+
+                        if (StringUtils.isNotEmpty(var.defaultValue)) { // has default value
+                            String defaultValue = var.defaultValue.substring(var.defaultValue.lastIndexOf('.') + 1);
+                            for (Map<String, Object> enumVars : (List<Map<String, Object>>) var.getAllowableValues().get("enumVars")) {
+                                if (defaultValue.equals(enumVars.get("name"))) {
+                                    // update default to use the string directly instead of enum string
+                                    var.defaultValue = (String) enumVars.get("value");
+                                }
+                            }
+                        }
+
+                        // add import for Set, HashSet
+                        cm.imports.add("Set");
+                        Map<String, String> importsSet = new HashMap<String, String>();
+                        importsSet.put("import", "java.util.Set");
+                        imports.add(importsSet);
+                        Map<String, String> importsHashSet = new HashMap<String, String>();
+                        importsHashSet.put("import", "java.util.HashSet");
+                        imports.add(importsHashSet);
+                    }
                 }
+
                 if (addImports) {
                     Map<String, String> imports2Classnames = new HashMap<String, String>() {{
                         put("JsonNullable", "org.openapitools.jackson.nullable.JsonNullable");

--- a/modules/openapi-generator/src/main/resources/Java/libraries/jersey2/pojo.mustache
+++ b/modules/openapi-generator/src/main/resources/Java/libraries/jersey2/pojo.mustache
@@ -101,7 +101,7 @@ public class {{classname}} {{#parent}}extends {{{parent}}} {{/parent}}{{#vendorE
 
   {{^isReadOnly}}
   {{#vendorExtensions.x-enum-as-string}}
-  private static final Set<String> {{{nameInSnakeCase}}}_VALUES = new HashSet<>(Arrays.asList(
+  public static final Set<String> {{{nameInSnakeCase}}}_VALUES = new HashSet<>(Arrays.asList(
     {{#allowableValues}}{{#enumVars}}{{{value}}}{{^-last}}, {{/-last}}{{/enumVars}}{{/allowableValues}}
   ));
 

--- a/modules/openapi-generator/src/main/resources/Java/libraries/jersey2/pojo.mustache
+++ b/modules/openapi-generator/src/main/resources/Java/libraries/jersey2/pojo.mustache
@@ -18,7 +18,9 @@ public class {{classname}} {{#parent}}extends {{{parent}}} {{/parent}}{{#vendorE
   {{#vars}}
     {{#isEnum}}
     {{^isContainer}}
+    {{^vendorExtensions.x-enum-as-string}}
 {{>modelInnerEnum}}
+    {{/vendorExtensions.x-enum-as-string}}
     {{/isContainer}}
     {{#isContainer}}
     {{#mostInnerItems}}
@@ -98,7 +100,19 @@ public class {{classname}} {{#parent}}extends {{{parent}}} {{/parent}}{{#vendorE
   {{#vars}}
 
   {{^isReadOnly}}
+  {{#vendorExtensions.x-enum-as-string}}
+  private static final Set<String> {{{nameInSnakeCase}}}_VALUES = new HashSet<>(Arrays.asList(
+    {{#allowableValues}}{{#enumVars}}{{{value}}}{{^-last}}, {{/-last}}{{/enumVars}}{{/allowableValues}}
+  ));
+
+  {{/vendorExtensions.x-enum-as-string}}
   public {{classname}} {{name}}({{{datatypeWithEnum}}} {{name}}) {
+    {{#vendorExtensions.x-enum-as-string}}
+    if (!{{{nameInSnakeCase}}}_VALUES.contains({{name}})) {
+      throw new IllegalArgumentException({{name}} + " is invalid. Possible values for {{name}}: " + String.join(", ", {{{nameInSnakeCase}}}_VALUES));
+    }
+
+    {{/vendorExtensions.x-enum-as-string}}
     {{#vendorExtensions.x-is-jackson-optional-nullable}}
     this.{{name}} = JsonNullable.<{{{datatypeWithEnum}}}>of({{name}});
     {{/vendorExtensions.x-is-jackson-optional-nullable}}
@@ -220,6 +234,12 @@ public class {{classname}} {{#parent}}extends {{{parent}}} {{/parent}}{{#vendorE
 
   {{^isReadOnly}}
   public void {{setter}}({{{datatypeWithEnum}}} {{name}}) {
+    {{#vendorExtensions.x-enum-as-string}}
+    if (!{{{nameInSnakeCase}}}_VALUES.contains({{name}})) {
+      throw new IllegalArgumentException({{name}} + " is invalid. Possible values for {{name}}: " + String.join(", ", {{{nameInSnakeCase}}}_VALUES));
+    }
+
+    {{/vendorExtensions.x-enum-as-string}}
     {{#vendorExtensions.x-is-jackson-optional-nullable}}
     this.{{name}} = JsonNullable.<{{{datatypeWithEnum}}}>of({{name}});
     {{/vendorExtensions.x-is-jackson-optional-nullable}}

--- a/modules/openapi-generator/src/test/resources/3_0/java/petstore-with-fake-endpoints-models-for-testing-with-http-signature.yaml
+++ b/modules/openapi-generator/src/test/resources/3_0/java/petstore-with-fake-endpoints-models-for-testing-with-http-signature.yaml
@@ -2087,6 +2087,12 @@ components:
           properties:
             name:
               type: string
+            pet_type:
+              x-enum-as-string: true
+              type: string
+              enum:
+              - ChildCat
+              default: ChildCat
     ArrayOfEnums:
       type: array
       items:

--- a/samples/client/petstore/csharp-netcore/OpenAPIClientCore/docs/ChildCat.md
+++ b/samples/client/petstore/csharp-netcore/OpenAPIClientCore/docs/ChildCat.md
@@ -4,6 +4,7 @@
 Name | Type | Description | Notes
 ------------ | ------------- | ------------- | -------------
 **Name** | **string** |  | [optional] 
+**PetType** | **string** |  | [optional] [default to PetTypeEnum.ChildCat]
 
 [[Back to Model list]](../README.md#documentation-for-models) [[Back to API list]](../README.md#documentation-for-api-endpoints) [[Back to README]](../README.md)
 

--- a/samples/client/petstore/csharp-netcore/OpenAPIClientCore/docs/ChildCatAllOf.md
+++ b/samples/client/petstore/csharp-netcore/OpenAPIClientCore/docs/ChildCatAllOf.md
@@ -4,6 +4,7 @@
 Name | Type | Description | Notes
 ------------ | ------------- | ------------- | -------------
 **Name** | **string** |  | [optional] 
+**PetType** | **string** |  | [optional] [default to PetTypeEnum.ChildCat]
 
 [[Back to Model list]](../README.md#documentation-for-models) [[Back to API list]](../README.md#documentation-for-api-endpoints) [[Back to README]](../README.md)
 

--- a/samples/client/petstore/csharp-netcore/OpenAPIClientCore/src/Org.OpenAPITools/Model/ChildCat.cs
+++ b/samples/client/petstore/csharp-netcore/OpenAPIClientCore/src/Org.OpenAPITools/Model/ChildCat.cs
@@ -32,6 +32,25 @@ namespace Org.OpenAPITools.Model
     public partial class ChildCat : ParentPet, IEquatable<ChildCat>, IValidatableObject
     {
         /// <summary>
+        /// Defines PetType
+        /// </summary>
+        [JsonConverter(typeof(StringEnumConverter))]
+        public enum PetTypeEnum
+        {
+            /// <summary>
+            /// Enum ChildCat for value: ChildCat
+            /// </summary>
+            [EnumMember(Value = "ChildCat")]
+            ChildCat = 1
+
+        }
+
+        /// <summary>
+        /// Gets or Sets PetType
+        /// </summary>
+        [DataMember(Name = "pet_type", EmitDefaultValue = false)]
+        public PetTypeEnum? PetType { get; set; }
+        /// <summary>
         /// Initializes a new instance of the <see cref="ChildCat" /> class.
         /// </summary>
         [JsonConstructorAttribute]
@@ -40,10 +59,11 @@ namespace Org.OpenAPITools.Model
         /// Initializes a new instance of the <see cref="ChildCat" /> class.
         /// </summary>
         /// <param name="name">name.</param>
-        /// <param name="petType">petType (required).</param>
-        public ChildCat(string name = default(string), string petType = default(string)) : base()
+        /// <param name="petType">petType (default to PetTypeEnum.ChildCat).</param>
+        public ChildCat(string name = default(string), PetTypeEnum? petType = PetTypeEnum.ChildCat) : base()
         {
             this.Name = name;
+            this.PetType = petType;
         }
 
         /// <summary>
@@ -62,6 +82,7 @@ namespace Org.OpenAPITools.Model
             sb.Append("class ChildCat {\n");
             sb.Append("  ").Append(base.ToString().Replace("\n", "\n  ")).Append("\n");
             sb.Append("  Name: ").Append(Name).Append("\n");
+            sb.Append("  PetType: ").Append(PetType).Append("\n");
             sb.Append("}\n");
             return sb.ToString();
         }
@@ -106,6 +127,7 @@ namespace Org.OpenAPITools.Model
                 int hashCode = base.GetHashCode();
                 if (this.Name != null)
                     hashCode = hashCode * 59 + this.Name.GetHashCode();
+                hashCode = hashCode * 59 + this.PetType.GetHashCode();
                 return hashCode;
             }
         }

--- a/samples/client/petstore/csharp-netcore/OpenAPIClientCore/src/Org.OpenAPITools/Model/ChildCatAllOf.cs
+++ b/samples/client/petstore/csharp-netcore/OpenAPIClientCore/src/Org.OpenAPITools/Model/ChildCatAllOf.cs
@@ -32,12 +32,33 @@ namespace Org.OpenAPITools.Model
     public partial class ChildCatAllOf : IEquatable<ChildCatAllOf>, IValidatableObject
     {
         /// <summary>
+        /// Defines PetType
+        /// </summary>
+        [JsonConverter(typeof(StringEnumConverter))]
+        public enum PetTypeEnum
+        {
+            /// <summary>
+            /// Enum ChildCat for value: ChildCat
+            /// </summary>
+            [EnumMember(Value = "ChildCat")]
+            ChildCat = 1
+
+        }
+
+        /// <summary>
+        /// Gets or Sets PetType
+        /// </summary>
+        [DataMember(Name = "pet_type", EmitDefaultValue = false)]
+        public PetTypeEnum? PetType { get; set; }
+        /// <summary>
         /// Initializes a new instance of the <see cref="ChildCatAllOf" /> class.
         /// </summary>
         /// <param name="name">name.</param>
-        public ChildCatAllOf(string name = default(string))
+        /// <param name="petType">petType (default to PetTypeEnum.ChildCat).</param>
+        public ChildCatAllOf(string name = default(string), PetTypeEnum? petType = PetTypeEnum.ChildCat)
         {
             this.Name = name;
+            this.PetType = petType;
         }
 
         /// <summary>
@@ -55,6 +76,7 @@ namespace Org.OpenAPITools.Model
             var sb = new StringBuilder();
             sb.Append("class ChildCatAllOf {\n");
             sb.Append("  Name: ").Append(Name).Append("\n");
+            sb.Append("  PetType: ").Append(PetType).Append("\n");
             sb.Append("}\n");
             return sb.ToString();
         }
@@ -99,6 +121,7 @@ namespace Org.OpenAPITools.Model
                 int hashCode = 41;
                 if (this.Name != null)
                     hashCode = hashCode * 59 + this.Name.GetHashCode();
+                hashCode = hashCode * 59 + this.PetType.GetHashCode();
                 return hashCode;
             }
         }

--- a/samples/openapi3/client/petstore/java/jersey2-java8/api/openapi.yaml
+++ b/samples/openapi3/client/petstore/java/jersey2-java8/api/openapi.yaml
@@ -2401,6 +2401,12 @@ components:
       properties:
         name:
           type: string
+        pet_type:
+          default: ChildCat
+          enum:
+          - ChildCat
+          type: string
+          x-enum-as-string: true
       type: object
   securitySchemes:
     petstore_auth:

--- a/samples/openapi3/client/petstore/java/jersey2-java8/docs/ChildCat.md
+++ b/samples/openapi3/client/petstore/java/jersey2-java8/docs/ChildCat.md
@@ -7,6 +7,15 @@
 Name | Type | Description | Notes
 ------------ | ------------- | ------------- | -------------
 **name** | **String** |  |  [optional]
+**petType** | [**String**](#String) |  |  [optional]
+
+
+
+## Enum: String
+
+Name | Value
+---- | -----
+CHILDCAT | &quot;ChildCat&quot;
 
 
 

--- a/samples/openapi3/client/petstore/java/jersey2-java8/docs/ChildCatAllOf.md
+++ b/samples/openapi3/client/petstore/java/jersey2-java8/docs/ChildCatAllOf.md
@@ -7,6 +7,15 @@
 Name | Type | Description | Notes
 ------------ | ------------- | ------------- | -------------
 **name** | **String** |  |  [optional]
+**petType** | [**String**](#String) |  |  [optional]
+
+
+
+## Enum: String
+
+Name | Value
+---- | -----
+CHILDCAT | &quot;ChildCat&quot;
 
 
 

--- a/samples/openapi3/client/petstore/java/jersey2-java8/src/main/java/org/openapitools/client/model/ChildCat.java
+++ b/samples/openapi3/client/petstore/java/jersey2-java8/src/main/java/org/openapitools/client/model/ChildCat.java
@@ -80,7 +80,7 @@ public class ChildCat extends ParentPet {
   }
 
 
-  private static final Set<String> PET_TYPE_VALUES = new HashSet<>(Arrays.asList(
+  public static final Set<String> PET_TYPE_VALUES = new HashSet<>(Arrays.asList(
     "ChildCat"
   ));
 

--- a/samples/openapi3/client/petstore/java/jersey2-java8/src/main/java/org/openapitools/client/model/ChildCat.java
+++ b/samples/openapi3/client/petstore/java/jersey2-java8/src/main/java/org/openapitools/client/model/ChildCat.java
@@ -32,6 +32,8 @@ import io.swagger.annotations.ApiModel;
 import io.swagger.annotations.ApiModelProperty;
 import org.openapitools.client.model.ChildCatAllOf;
 import org.openapitools.client.model.ParentPet;
+import java.util.Set;
+import java.util.HashSet;
 import com.fasterxml.jackson.annotation.JsonPropertyOrder;
 import org.openapitools.client.JSON;
 
@@ -40,7 +42,8 @@ import org.openapitools.client.JSON;
  * ChildCat
  */
 @JsonPropertyOrder({
-  ChildCat.JSON_PROPERTY_NAME
+  ChildCat.JSON_PROPERTY_NAME,
+  ChildCat.JSON_PROPERTY_PET_TYPE
 })
 @javax.annotation.Generated(value = "org.openapitools.codegen.languages.JavaClientCodegen")
 @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.EXISTING_PROPERTY, property = "pet_type", visible = true)
@@ -48,6 +51,9 @@ import org.openapitools.client.JSON;
 public class ChildCat extends ParentPet {
   public static final String JSON_PROPERTY_NAME = "name";
   private String name;
+
+  public static final String JSON_PROPERTY_PET_TYPE = "pet_type";
+  private String petType = "ChildCat";
 
 
   public ChildCat name(String name) {
@@ -71,6 +77,42 @@ public class ChildCat extends ParentPet {
 
   public void setName(String name) {
     this.name = name;
+  }
+
+
+  private static final Set<String> PET_TYPE_VALUES = new HashSet<>(Arrays.asList(
+    "ChildCat"
+  ));
+
+  public ChildCat petType(String petType) {
+    if (!PET_TYPE_VALUES.contains(petType)) {
+      throw new IllegalArgumentException(petType + " is invalid. Possible values for petType: " + String.join(", ", PET_TYPE_VALUES));
+    }
+
+    this.petType = petType;
+    return this;
+  }
+
+   /**
+   * Get petType
+   * @return petType
+  **/
+  @javax.annotation.Nullable
+  @ApiModelProperty(value = "")
+  @JsonProperty(JSON_PROPERTY_PET_TYPE)
+  @JsonInclude(value = JsonInclude.Include.USE_DEFAULTS)
+
+  public String getPetType() {
+    return petType;
+  }
+
+
+  public void setPetType(String petType) {
+    if (!PET_TYPE_VALUES.contains(petType)) {
+      throw new IllegalArgumentException(petType + " is invalid. Possible values for petType: " + String.join(", ", PET_TYPE_VALUES));
+    }
+
+    this.petType = petType;
   }
 
   /**
@@ -123,14 +165,15 @@ public class ChildCat extends ParentPet {
       return false;
     }
     ChildCat childCat = (ChildCat) o;
-    return Objects.equals(this.name, childCat.name)&&
+    return Objects.equals(this.name, childCat.name) &&
+        Objects.equals(this.petType, childCat.petType)&&
         Objects.equals(this.additionalProperties, childCat.additionalProperties) &&
         super.equals(o);
   }
 
   @Override
   public int hashCode() {
-    return Objects.hash(name, super.hashCode(), additionalProperties);
+    return Objects.hash(name, petType, super.hashCode(), additionalProperties);
   }
 
 
@@ -140,6 +183,7 @@ public class ChildCat extends ParentPet {
     sb.append("class ChildCat {\n");
     sb.append("    ").append(toIndentedString(super.toString())).append("\n");
     sb.append("    name: ").append(toIndentedString(name)).append("\n");
+    sb.append("    petType: ").append(toIndentedString(petType)).append("\n");
     sb.append("    additionalProperties: ").append(toIndentedString(additionalProperties)).append("\n");
     sb.append("}");
     return sb.toString();

--- a/samples/openapi3/client/petstore/java/jersey2-java8/src/main/java/org/openapitools/client/model/ChildCatAllOf.java
+++ b/samples/openapi3/client/petstore/java/jersey2-java8/src/main/java/org/openapitools/client/model/ChildCatAllOf.java
@@ -70,7 +70,7 @@ public class ChildCatAllOf {
   }
 
 
-  private static final Set<String> PET_TYPE_VALUES = new HashSet<>(Arrays.asList(
+  public static final Set<String> PET_TYPE_VALUES = new HashSet<>(Arrays.asList(
     "ChildCat"
   ));
 

--- a/samples/openapi3/client/petstore/java/jersey2-java8/src/main/java/org/openapitools/client/model/ChildCatAllOf.java
+++ b/samples/openapi3/client/petstore/java/jersey2-java8/src/main/java/org/openapitools/client/model/ChildCatAllOf.java
@@ -24,6 +24,8 @@ import com.fasterxml.jackson.annotation.JsonTypeName;
 import com.fasterxml.jackson.annotation.JsonValue;
 import io.swagger.annotations.ApiModel;
 import io.swagger.annotations.ApiModelProperty;
+import java.util.Set;
+import java.util.HashSet;
 import com.fasterxml.jackson.annotation.JsonPropertyOrder;
 import org.openapitools.client.JSON;
 
@@ -32,12 +34,16 @@ import org.openapitools.client.JSON;
  * ChildCatAllOf
  */
 @JsonPropertyOrder({
-  ChildCatAllOf.JSON_PROPERTY_NAME
+  ChildCatAllOf.JSON_PROPERTY_NAME,
+  ChildCatAllOf.JSON_PROPERTY_PET_TYPE
 })
 @javax.annotation.Generated(value = "org.openapitools.codegen.languages.JavaClientCodegen")
 public class ChildCatAllOf {
   public static final String JSON_PROPERTY_NAME = "name";
   private String name;
+
+  public static final String JSON_PROPERTY_PET_TYPE = "pet_type";
+  private String petType = "ChildCat";
 
 
   public ChildCatAllOf name(String name) {
@@ -64,6 +70,42 @@ public class ChildCatAllOf {
   }
 
 
+  private static final Set<String> PET_TYPE_VALUES = new HashSet<>(Arrays.asList(
+    "ChildCat"
+  ));
+
+  public ChildCatAllOf petType(String petType) {
+    if (!PET_TYPE_VALUES.contains(petType)) {
+      throw new IllegalArgumentException(petType + " is invalid. Possible values for petType: " + String.join(", ", PET_TYPE_VALUES));
+    }
+
+    this.petType = petType;
+    return this;
+  }
+
+   /**
+   * Get petType
+   * @return petType
+  **/
+  @javax.annotation.Nullable
+  @ApiModelProperty(value = "")
+  @JsonProperty(JSON_PROPERTY_PET_TYPE)
+  @JsonInclude(value = JsonInclude.Include.USE_DEFAULTS)
+
+  public String getPetType() {
+    return petType;
+  }
+
+
+  public void setPetType(String petType) {
+    if (!PET_TYPE_VALUES.contains(petType)) {
+      throw new IllegalArgumentException(petType + " is invalid. Possible values for petType: " + String.join(", ", PET_TYPE_VALUES));
+    }
+
+    this.petType = petType;
+  }
+
+
   /**
    * Return true if this ChildCat_allOf object is equal to o.
    */
@@ -76,12 +118,13 @@ public class ChildCatAllOf {
       return false;
     }
     ChildCatAllOf childCatAllOf = (ChildCatAllOf) o;
-    return Objects.equals(this.name, childCatAllOf.name);
+    return Objects.equals(this.name, childCatAllOf.name) &&
+        Objects.equals(this.petType, childCatAllOf.petType);
   }
 
   @Override
   public int hashCode() {
-    return Objects.hash(name);
+    return Objects.hash(name, petType);
   }
 
 
@@ -90,6 +133,7 @@ public class ChildCatAllOf {
     StringBuilder sb = new StringBuilder();
     sb.append("class ChildCatAllOf {\n");
     sb.append("    name: ").append(toIndentedString(name)).append("\n");
+    sb.append("    petType: ").append(toIndentedString(petType)).append("\n");
     sb.append("}");
     return sb.toString();
   }

--- a/samples/openapi3/client/petstore/java/jersey2-java8/src/test/java/org/openapitools/client/JSONComposedSchemaTest.java
+++ b/samples/openapi3/client/petstore/java/jersey2-java8/src/test/java/org/openapitools/client/JSONComposedSchemaTest.java
@@ -67,6 +67,18 @@ public class JSONComposedSchemaTest {
     }
 
     /**
+     * Test to ensure the setter will throw IllegalArgumentException
+     */
+    @Test(expected = IllegalArgumentException.class)
+    public void testEnumDiscriminator() throws Exception {
+        ChildCat cc = new ChildCat();
+        cc.setPetType("ChildCat");
+        assertEquals("ChildCat", cc.getPetType());
+
+        cc.setPetType("WrongValue");
+    }
+
+    /**
      * Test to ensure the getter will throw ClassCastException
      */
     @Test(expected = ClassCastException.class)

